### PR TITLE
BUG: Properly handle lists for .mask

### LINKED ISSUE #21891: Incorrect assumption in Series.mask()
#### Code Sample, a copy-pastable example if possible

```python
pd.Series([1, 2]).where([True, False])
# 0    1.0
# 1    NaN
# dtype: float64
pd.Series([1, 2]).mask([True, False])
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
#   File "/usr/local/lib/python3.7/site-packages/pandas/core/generic.py", line 7792, in mask
#     return self.where(~cond, other=other, inplace=inplace, axis=axis,
# TypeError: bad operand type for unary ~: 'list'
```
#### Problem description

With the same `cond`, `Series.where(cond)` succeeded but `Series.mask(cond)` failed.
It's because  `Series.mask()` assumes `cond` has `__invert__()` while it's not always ture.

https://github.com/pandas-dev/pandas/blob/365eac4d27353693a2b758c87fc4bceabb74817f/pandas/core/generic.py#L7944-L7946
#### Expected Output
`Series.mask(cond)` succeeds.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Darwin
OS-release: 17.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: zh_CN.UTF-8
LOCALE: zh_CN.UTF-8

pandas: 0.23.3
pytest: None
pip: 10.0.1
setuptools: 39.2.0
Cython: None
numpy: 1.14.5
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.

Closes #21891.

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -7941,6 +7941,10 @@ def mask(self, cond, other=np.nan, inplace=False, axis=None, level=None,
         inplace = validate_bool_kwarg(inplace, 'inplace')
         cond = com._apply_if_callable(cond, self)
 
+        # see gh-21891
```
Comment (jreback) [inline pandas/core/generic.py]: I would do this in ``.where`` itself

File: pandas/tests/frame/test_indexing.py [hunk 2]
```diff
@@ -3010,6 +3010,14 @@ def test_mask_callable(self):
         tm.assert_frame_equal(result,
                               (df + 2).mask((df + 2) > 8, (df + 2) + 10))
 
+    def test_mask_list(self):
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: can you pas a bunch of list-likes in a fixture (the old box argument)

File: pandas/tests/series/indexing/test_boolean.py [hunk 3]
```diff
@@ -582,6 +582,15 @@ def test_where_dt_tz_values(tz_naive_fixture):
     assert_series_equal(exp, result)
 
 
+def test_where_list():
```
Comment (jreback) [inline pandas/tests/series/indexing/test_boolean.py]: same

Comment (jreback) [reply pandas/tests/series/indexing/test_boolean.py]: can you test where as well with array-likes for the cond

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (gfyoung) [reply pandas/core/generic.py]: Do what exactly in `.where` ?

Comment (jreback) [reply pandas/core/generic.py]: oh i c we are using ``~`` here, ok then.

File: pandas/tests/series/indexing/test_boolean.py (hunk 3 above)
Comment (gfyoung) [reply pandas/tests/series/indexing/test_boolean.py]: > can you test where as well with array-likes for the cond

That's tested pretty heavily for `DataFrame` and `Series` already in fact.

File: pandas/tests/frame/test_indexing.py (hunk 2 above)
Comment (gfyoung) [reply pandas/tests/frame/test_indexing.py]: Old box argument?  Sorry, I'm not sure what you mean here (or I forgot what it meant if you have mentioned it to me before).

File: pandas/tests/series/indexing/test_boolean.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/series/indexing/test_boolean.py]: ok can u integrate this test in a related one then?

Comment (gfyoung) [reply pandas/tests/series/indexing/test_boolean.py]: I thought about that, but I was more comfortable leaving it as a separate test semantically speaking.  After all, it is more atomic that way IMO.

Comment (jreback) [reply pandas/tests/series/indexing/test_boolean.py]: right but it is not clear not where things are tested. we are apparently testing list-likes (ex list) somewhere else, so this is confusing to a new reader and is just future technical debt.

Comment (gfyoung) [reply pandas/tests/series/indexing/test_boolean.py]: Fair enough.  Integrated.

File: pandas/core/generic.py (hunk 1 above)
Comment (holymonson) [inline pandas/core/generic.py]: `cond ` could be a `callable `, this patch seems wouldn't work.

**Review by holymonson — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (gfyoung) [reply pandas/core/generic.py]: Yes, it would.  Because `com._apply_if_callable` (right above) returns an `ndarray`.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @gfyoung

**Pull request merged**